### ISSUE TITLE
State is updated with new yaml content even if apply fails

### ISSUE THREAD

**Issue by nicolst** [alekc/terraform-provider-kubectl#60]
When rolling out a change in a manifest, the state file is updated with the new yaml body even if the apply failed, e.g. due to expired credentials etc. If you run plan after this failed apply, it will report no changes, even though the actual in-cluster resources have not been updated.

More specifically I tried to apply a plan which was planned the day before, which failed because of expired credentials (for Google Kubernetes Engine). Subsequent plan showed no changes, even though the in-cluster yaml had not changed. Rolling back the statefile to the previous version and plan/apply-ing again solved the issue.

**Comment by nicolst** [comment]
I see now that this is a duplicate of existing issue in gavinbunney's repo https://github.com/gavinbunney/terraform-provider-kubectl/issues/265

**Comment by alekc** [comment]
Yes, I've seen that issue as well. The diff checking mechanism is pretty much "broken" at the moment, it generates a lot of false positives (yaml_incluster has been changed xxx), and lack of positive on the real issue as yours. I am looking into it, but it's going to take some time because it would require significant refactoring.

**Comment by robpearce-flux** [comment]
believe we're running into this issue also, the quote on https://developer.hashicorp.com/terraform/plugin/framework/diagnostics#how-errors-affect-state suggests "When returning error diagnostics, we recommend resetting the state in the response to the prior state available in the configuration." Is that the refactoring work you're describing ?

**Comment by alekc** [comment]
Fixed by PR #320 (`e009949`). The framework Update path now captures req.State into priorData at entry and explicitly restores it on every error branch via resp.State.Set before returning. TestUpdate_PreservesStateOnApplyError pins the regression contract.
